File: src/widgets/spriteset/spriteseteditor.cpp
#include <QMenu>
#include <QAction>
#include <QHBoxLayout>
#include <QScrollArea>
#include <QSplitter>
#include <QGraphicsItem>
#include <QGraphicsPixmapItem>
#include <QGraphicsView>
#include <QGraphicsScene>
#include <QLineEdit>
#include <QToolButton>
#include <QFileInfo>

#include "widgets/spriteset/spriteseteditor.h"
#include "ui_spriteseteditor.h"
#include "widgets/palettewidget.h"
#include "widgets/imagechooser.h"
#include "mainwindow.h"
#include "widgets/spriteset/ssdirectionview.h"
#include "formats/spriteset.h"
#include "widgets/sphereeditor.h"

SpritesetEditor::SpritesetEditor(QWidget *parent): SphereEditor(parent), ui(new Ui::SpritesetEditor) {
	ui->setupUi(this);
	m_type = SphereEditor::SpritesetEditor;
	setContextMenuPolicy(Qt::CustomContextMenu);
	connect(this, SIGNAL(customContextMenuRequested(const QPoint&)),
			this, SLOT(showContextMenu(const QPoint&)));
	setBackgroundRole(QPalette::Shadow);
	ui->dirsLayout->setContentsMargins(8,4,4,4);
	ui->dirsLayout->setSpacing(0);
	PaletteWidget* pw = new PaletteWidget(this);

	ui->palette->setWidget(pw);
	ui->palette->setAlignment(Qt::AlignTop|Qt::AlignLeft);
	QPalette pal = palette();
	pal.setColor(QPalette::Background, QColor(40,40,40));

	m_spriteset = new Spriteset(this);
	m_images = new ImageChooser(true, this);
	ui->ssImages->setWidget(m_images);
	ui->ssImages->setAlignment(Qt::AlignLeft|Qt::AlignTop);

	int mainWidth = MainWindow::instance()->width();
	ui->ssViewSplitter->setStretchFactor(0,4);
	ui->ssViewSplitter->setSizes(QList<int>({mainWidth-164,164}));
	ui->choosers->setSizes(QList<int>({200,150,200}));

	ui->animView->setBackgroundBrush(QBrush(QPixmap(":/icons/transparency-bg.png")));
	ui->dirsLayout->addStretch(5);
}

SpritesetEditor::~SpritesetEditor() {
	delete ui;
	disconnect(this, SIGNAL(customContextMenuRequested(const QPoint&)),
			   this, SLOT(showContextMenu(const QPoint&)));
}

bool SpritesetEditor::attach(Spriteset* spriteset) {
	m_spriteset = spriteset;
	m_images->setImages(m_spriteset->getImages());
	if(m_spriteset->numImages() == 0) return true;

	const QPixmap animPixmap = QPixmap::fromImage(*m_spriteset->getImage(0));
	QGraphicsPixmapItem* animFrame = new QGraphicsPixmapItem(animPixmap);
	ui->animView->setScene(new QGraphicsScene(ui->animView));
	animFrame->setScale(2);
	ui->animView->scene()->addItem(animFrame);
	int num_directions = m_spriteset->numDirections();
	ui->animDirChoose->clear();
	for(int d = 0; d < num_directions; d++) {
		Spriteset::SSDirection* direction = m_spriteset->getDirection(d);
		ui->animDirChoose->addItem(direction->name);
		attachDirection(d);
	}

	return true;
}

Spriteset *SpritesetEditor::getSpriteset() {
	return m_spriteset;
}

bool SpritesetEditor::openFile(QString filename) {
	bool success = m_spriteset->open(filename);
	if(!success) return false;
	return attach(m_spriteset);
}

void SpritesetEditor::addDirection(QString name, int numFrames) {
	m_spriteset->addDirection(name, numFrames);
	SSDirectionView* dirView = new SSDirectionView(m_spriteset, m_spriteset->numDirections()-1, this);
	ui->dirsContainer->addWidget(dirView);
}

void SpritesetEditor::attachDirection(int index) {
	SSDirectionView* dirView = new SSDirectionView(m_spriteset, index, this);
	ui->dirsContainer->addWidget(dirView);
}

void SpritesetEditor::showContextMenu(const QPoint &pos) {
	QMenu menu(this);
	menu.addAction("Add direction");
	QAction* result = menu.exec(mapToGlobal(pos));
	if(!result) return;
	if(result->text() == "Add direction") {
		addDirection("",1);
	}
}

void SpritesetEditor::on_animDirChoose_currentIndexChanged(int index) {

}

File: widgets/spriteset/spriteseteditor.py
from PySide6.QtCore import Qt, Slot, QObject
from PySide6.QtGui import QPixmap
from PySide6.QtWidgets import QWidget, QGraphicsPixmapItem, QGraphicsScene

from formats.spriteset import SphereSpriteset
from ..sphereeditor import SphereEditor, SphereFile
from .imagechooser import ImageChooser
from .directionview import DirectionView

from ui.ui_spriteseteditor import Ui_SpritesetEditor

class SpritesetEditor(SphereEditor):
	ui: Ui_SpritesetEditor
	loadedSpriteset: SphereSpriteset
	imagesScene: QGraphicsScene
	directionScene: QGraphicsScene
	directionViews: list[DirectionView]
	imageChooser: ImageChooser

	@staticmethod
	def openAndAttach(parent: QWidget, filePath:str):
		rss = SphereSpriteset(filePath)
		rss.open()
		editor = SpritesetEditor(parent)
		editor.attachSpriteset(rss)
		return editor

	def __init__(self, parent: QWidget = None):
		super().__init__(parent)
		self.ui = Ui_SpritesetEditor()
		self.ui.setupUi(self)
		self.directionViews = []
		self.editorType = SphereFile.Spriteset
		self.loadedSpriteset = None
		self.directionScene = QGraphicsScene(self.ui.animView)
		self.ui.animView.setScene(self.directionScene)
		self.imagesScene = QGraphicsScene(self.ui.ssImages)
		self.ui.animDirChoose.currentIndexChanged.connect(self.onDirectionChanged)
		self.imageChooser = ImageChooser(self, True)
		self.ui.ssImages.setWidget(self.imageChooser)
		self.ui.ssImages.setAlignment(Qt.AlignmentFlag.AlignLeft|Qt.AlignmentFlag.AlignTop)
		self._adjustSizes()

	def _adjustSizes(self):
		self.ui.ssViewSplitter.setStretchFactor(0,4)
		self.ui.ssViewSplitter.setSizes((self.window().width()-164,164))
		self.ui.choosers.setSizes((200,150,200))

	def setDirection(self, d:int):
		self.directionScene.clear()
		image = self.loadedSpriteset.images[self.loadedSpriteset.directions[d].frames[0].imageIndex]
		item = QGraphicsPixmapItem(QPixmap.fromImage(image)) # the internal C++ object is deleted when scene.clear() is called so we have to recreate it
		item.setScale(2)
		self.directionScene.addItem(item)


	def attachSpriteset(self, spriteset:SphereSpriteset):
		self.ui.animDirChoose.clear()
		self.filePath = spriteset.filePath
		for d in range(len(spriteset.directions)):
			self.ui.animDirChoose.addItem(spriteset.directions[d].name)
			view = DirectionView(self, spriteset, d, self.undoStack)
			# view.addFrameButton.clicked.connect(lambda: self.onAddDirectionFrameClicked(d))
			self.directionViews.append(view)
			self.ui.dirsContainer.addWidget(view)


		self.imageChooser.images = spriteset.images
		self.loadedSpriteset = spriteset
		self.setDirection(0)

	@Slot(int)
	def onDirectionChanged(self, d:int):
		if self.loadedSpriteset is None or d < 0 or d >= len(self.loadedSpriteset.directions):
			return
		self.setDirection(d)
